File: Sources/ComponentLibrary/Components/DSButton/DSButton.swift
import SwiftUI
import DesignTokens

/// A customizable button component with multiple styles and sizes
///
/// `DSButton` provides a consistent button interface across your app with
/// support for different visual styles, sizes, and states.
///
/// ## Example
///
/// ```swift
/// DSButton(
///     title: "Submit",
///     style: .primary,
///     size: .medium
/// ) {
///     print("Button tapped")
/// }
/// ```
public struct DSButton: View {
    // MARK: - Properties

    private let title: String
    private let style: DSButtonStyle
    private let size: DSButtonSize
    private let isDisabled: Bool
    private let isLoading: Bool
    private let action: () -> Void

    // MARK: - Initialization

    /// Creates a new button with the specified configuration
    ///
    /// - Parameters:
    ///   - title: The text to display on the button
    ///   - style: The visual style of the button (default: `.primary`)
    ///   - size: The size of the button (default: `.medium`)
    ///   - isDisabled: Whether the button is disabled (default: `false`)
    ///   - isLoading: Whether to show a loading indicator (default: `false`)
    ///   - action: The action to perform when the button is tapped
    public init(
        title: String,
        style: DSButtonStyle = .primary,
        size: DSButtonSize = .medium,
        isDisabled: Bool = false,
        isLoading: Bool = false,
        action: @escaping () -> Void
    ) {
        self.title = title
        self.style = style
        self.size = size
        self.isDisabled = isDisabled
        self.isLoading = isLoading
        self.action = action
    }

    // MARK: - Body

    public var body: some View {
        Button(action: action) {
            HStack(spacing: 8) {
                if isLoading {
                    ProgressView()
                        .progressViewStyle(CircularProgressViewStyle(tint: style.foregroundColor))
                }

                Text(title)
                    .font(size.font)
                    .fontWeight(.semibold)
            }
            .foregroundColor(style.foregroundColor)
            .padding(.horizontal, size.horizontalPadding)
            .padding(.vertical, size.verticalPadding)
            .background(style.backgroundColor)
            .cornerRadius(8)
            .overlay(
                RoundedRectangle(cornerRadius: 8)
                    .stroke(style.borderColor, lineWidth: style.borderWidth)
            )
        }
        .disabled(isDisabled || isLoading)
        .opacity(isDisabled ? 0.5 : 1.0)
    }
}

// MARK: - Button Style

/// Visual styles for DSButton
public enum DSButtonStyle {
    case primary
    case secondary
    case outline
    case ghost
    case destructive

    var backgroundColor: Color {
        switch self {
        case .primary:
            return Color(hex: "#2563eb") // primary-600
        case .secondary:
            return Color(hex: "#f5f5f5") // neutral-100
        case .outline, .ghost:
            return .clear
        case .destructive:
            return Color(hex: "#dc2626") // error-600
        }
    }

    var foregroundColor: Color {
        switch self {
        case .primary, .destructive:
            return .white
        case .secondary:
            return Color(hex: "#171717") // neutral-900
        case .outline, .ghost:
            return Color(hex: "#2563eb") // primary-600
        }
    }

    var borderColor: Color {
        switch self {
        case .outline:
            return Color(hex: "#2563eb") // primary-600
        case .primary, .secondary, .ghost, .destructive:
            return .clear
        }
    }

    var borderWidth: CGFloat {
        switch self {
        case .outline:
            return 1
        case .primary, .secondary, .ghost, .destructive:
            return 0
        }
    }
}

// MARK: - Button Size

/// Size variants for DSButton
public enum DSButtonSize {
    case small
    case medium
    case large

    var font: Font {
        switch self {
        case .small:
            return .system(size: 14)
        case .medium:
            return .system(size: 16)
        case .large:
            return .system(size: 18)
        }
    }

    var horizontalPadding: CGFloat {
        switch self {
        case .small:
            return 12
        case .medium:
            return 16
        case .large:
            return 20
        }
    }

    var verticalPadding: CGFloat {
        switch self {
        case .small:
            return 8
        case .medium:
            return 12
        case .large:
            return 16
        }
    }
}

// MARK: - Color Extension

extension Color {
    init(hex: String) {
        let hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&int)
        let a, r, g, b: UInt64
        switch hex.count {
        case 3: // RGB (12-bit)
            (a, r, g, b) = (255, (int >> 8) * 17, (int >> 4 & 0xF) * 17, (int & 0xF) * 17)
        case 6: // RGB (24-bit)
            (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
        case 8: // ARGB (32-bit)
            (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
        default:
            (a, r, g, b) = (1, 1, 1, 0)
        }

        self.init(
            .sRGB,
            red: Double(r) / 255,
            green: Double(g) / 255,
            blue:  Double(b) / 255,
            opacity: Double(a) / 255
        )
    }
}

// MARK: - Previews

#Preview("Button Styles") {
    VStack(spacing: 16) {
        DSButton(title: "Primary", style: .primary) {}
        DSButton(title: "Secondary", style: .secondary) {}
        DSButton(title: "Outline", style: .outline) {}
        DSButton(title: "Ghost", style: .ghost) {}
        DSButton(title: "Destructive", style: .destructive) {}
    }
    .padding()
}

#Preview("Button Sizes") {
    VStack(spacing: 16) {
        DSButton(title: "Small", size: .small) {}
        DSButton(title: "Medium", size: .medium) {}
        DSButton(title: "Large", size: .large) {}
    }
    .padding()
}

#Preview("Button States") {
    VStack(spacing: 16) {
        DSButton(title: "Normal") {}
        DSButton(title: "Disabled", isDisabled: true) {}
        DSButton(title: "Loading", isLoading: true) {}
    }
    .padding()
}

File: component-library/src/main/java/com/aaanandan/componentlibrary/components/DSButton.kt
package com.aaanandan.componentlibrary.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * A customizable button component with multiple styles and sizes
 *
 * DSButton provides a consistent button interface across your app with
 * support for different visual styles, sizes, and states.
 *
 * @param text The text to display on the button
 * @param onClick Callback when the button is clicked
 * @param modifier Modifier to be applied to the button
 * @param style The visual style of the button (default: Primary)
 * @param size The size of the button (default: Medium)
 * @param enabled Whether the button is enabled (default: true)
 * @param loading Whether to show a loading indicator (default: false)
 *
 * Example usage:
 * ```
 * DSButton(
 *     text = "Submit",
 *     style = DSButtonStyle.Primary,
 *     size = DSButtonSize.Medium,
 *     onClick = { /* Handle click */ }
 * )
 * ```
 */
@Composable
fun DSButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    style: DSButtonStyle = DSButtonStyle.Primary,
    size: DSButtonSize = DSButtonSize.Medium,
    enabled: Boolean = true,
    loading: Boolean = false
) {
    val colors = style.getButtonColors()
    val contentPadding = size.getContentPadding()
    val fontSize = size.getFontSize()

    when (style) {
        is DSButtonStyle.Outline -> {
            OutlinedButton(
                onClick = onClick,
                modifier = modifier,
                enabled = enabled && !loading,
                colors = colors,
                contentPadding = contentPadding,
                border = BorderStroke(1.dp, style.borderColor)
            ) {
                ButtonContent(text, fontSize, loading)
            }
        }
        is DSButtonStyle.Ghost -> {
            TextButton(
                onClick = onClick,
                modifier = modifier,
                enabled = enabled && !loading,
                colors = colors,
                contentPadding = contentPadding
            ) {
                ButtonContent(text, fontSize, loading)
            }
        }
        else -> {
            Button(
                onClick = onClick,
                modifier = modifier,
                enabled = enabled && !loading,
                colors = colors,
                contentPadding = contentPadding
            ) {
                ButtonContent(text, fontSize, loading)
            }
        }
    }
}

@Composable
private fun ButtonContent(
    text: String,
    fontSize: androidx.compose.ui.unit.TextUnit,
    loading: Boolean
) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        if (loading) {
            CircularProgressIndicator(
                modifier = Modifier.size(16.dp),
                strokeWidth = 2.dp
            )
        }
        Text(
            text = text,
            fontSize = fontSize
        )
    }
}

/**
 * Visual styles for DSButton
 */
sealed class DSButtonStyle {
    object Primary : DSButtonStyle()
    object Secondary : DSButtonStyle()
    data class Outline(val borderColor: Color = Color(0xFF2563EB)) : DSButtonStyle()
    object Ghost : DSButtonStyle()
    object Destructive : DSButtonStyle()

    @Composable
    fun getButtonColors(): ButtonColors {
        return when (this) {
            is Primary -> ButtonDefaults.buttonColors(
                containerColor = Color(0xFF2563EB), // primary-600
                contentColor = Color.White
            )
            is Secondary -> ButtonDefaults.buttonColors(
                containerColor = Color(0xFFF5F5F5), // neutral-100
                contentColor = Color(0xFF171717) // neutral-900
            )
            is Outline -> ButtonDefaults.outlinedButtonColors(
                containerColor = Color.Transparent,
                contentColor = Color(0xFF2563EB) // primary-600
            )
            is Ghost -> ButtonDefaults.textButtonColors(
                containerColor = Color.Transparent,
                contentColor = Color(0xFF2563EB) // primary-600
            )
            is Destructive -> ButtonDefaults.buttonColors(
                containerColor = Color(0xFFDC2626), // error-600
                contentColor = Color.White
            )
        }
    }
}

/**
 * Size variants for DSButton
 */
enum class DSButtonSize {
    Small,
    Medium,
    Large;

    fun getContentPadding(): PaddingValues {
        return when (this) {
            Small -> PaddingValues(horizontal = 12.dp, vertical = 8.dp)
            Medium -> PaddingValues(horizontal = 16.dp, vertical = 12.dp)
            Large -> PaddingValues(horizontal = 20.dp, vertical = 16.dp)
        }
    }

    fun getFontSize(): androidx.compose.ui.unit.TextUnit {
        return when (this) {
            Small -> 14.sp
            Medium -> 16.sp
            Large -> 18.sp
        }
    }
}

// ===== Previews =====

@Preview(name = "Button Styles", showBackground = true)
@Composable
private fun PreviewButtonStyles() {
    Column(
        modifier = Modifier.padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        DSButton(text = "Primary", style = DSButtonStyle.Primary, onClick = {})
        DSButton(text = "Secondary", style = DSButtonStyle.Secondary, onClick = {})
        DSButton(text = "Outline", style = DSButtonStyle.Outline(), onClick = {})
        DSButton(text = "Ghost", style = DSButtonStyle.Ghost, onClick = {})
        DSButton(text = "Destructive", style = DSButtonStyle.Destructive, onClick = {})
    }
}

@Preview(name = "Button Sizes", showBackground = true)
@Composable
private fun PreviewButtonSizes() {
    Column(
        modifier = Modifier.padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        DSButton(text = "Small", size = DSButtonSize.Small, onClick = {})
        DSButton(text = "Medium", size = DSButtonSize.Medium, onClick = {})
        DSButton(text = "Large", size = DSButtonSize.Large, onClick = {})
    }
}

@Preview(name = "Button States", showBackground = true)
@Composable
private fun PreviewButtonStates() {
    Column(
        modifier = Modifier.padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        DSButton(text = "Normal", onClick = {})
        DSButton(text = "Disabled", enabled = false, onClick = {})
        DSButton(text = "Loading", loading = true, onClick = {})
    }
}
